### PR TITLE
Move FS commands to memory registry

### PR DESCRIPTION
**Pull request by dsemenets**

### CODE REVIEW

File: tee_supplicant/src/ree_fs.c [hunk 1]
```diff
@@ -47,37 +94,46 @@ static int tee_fs_open(size_t num_params, struct tee_param *params,
...
+				goto out;
+			}
+			rc = fs_mkdir(dir);
+			if (rc < 0) {
+				rc = TEEC_ERROR_GENERIC;
+				goto out;
+			}
+			strcpy(path, REE_FS_MP);
+			strncat(path, name, PATH_MAX);
+			fd = internal_open(path, flags);
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/ree_fs.c]: We should handle /1/2/3/file here. Or if it's impossible - just check path to be ./file or ./1/file. 
If directory is well known why don't create it  on init?

Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: optee doesn't use directories. This is necessary to create directory for secure storage. Create code which will not be used is not good idea.

File: tee_supplicant/src/ree_fs.c [hunk 2]
```diff
@@ -21,12 +21,59 @@ LOG_MODULE_REGISTER(ree_fs);
 #define REE_FS_PATHLEN sizeof(REE_FS_MP)
 #define REE_FS_PATH_MAX (PATH_MAX + REE_FS_PATHLEN)
 
+static int internal_open(char *path, fs_mode_t flags)
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: This patch does more that commit description says. Looks like you are factoring out some code, but it is hard to tell what your intention is, because commit description say nothing about this.

File: tee_supplicant/src/ree_fs.c [hunk 3]
```diff
@@ -21,12 +21,59 @@ LOG_MODULE_REGISTER(ree_fs);
...
+	int fd, rc;
+	struct fs_file_t *file;
+
+	file = k_malloc(sizeof(*file));
+	if (!file) {
+		return -ENOMEM;
+	}
+
+	fs_file_t_init(file);
+	fd = z_alloc_fd(file, NULL);
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: Looks like `z_alloc_fd` is some internal API which will not be available to user mode. Why it is used?

File: tee_supplicant/src/ree_fs.c [hunk 4]
```diff
@@ -21,12 +21,59 @@ LOG_MODULE_REGISTER(ree_fs);
...
+	return fd;
+free:
+	if (fd >= 0) {
+		z_free_fd(fd);
+	}
+	k_free(file);
+	return rc;
+}
+
+static char *dname(char *path)
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: This will not handle some cases like `//////` or `/a/../b` or `/./`

You need to either prove that those cases are impossible or account for all of them

File: tee_supplicant/src/ree_fs.c [hunk 5]
```diff
@@ -150,7 +206,6 @@ static int tee_fs_read(size_t num_params, struct tee_param *params)
 		return TEEC_ERROR_ITEM_NOT_FOUND;
 	}
 
-	/*TODO: handle sz < len */
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:153]: Was this handled somehow?

File: tee_supplicant/src/ree_fs.c [hunk 6]
```diff
@@ -200,7 +255,6 @@ static int tee_fs_write(size_t num_params, struct tee_param *params)
 		return TEEC_ERROR_ITEM_NOT_FOUND;
 	}
 
-	/*TODO: handle case of partially written buffer */
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:203]: Was this handled somehow?

File: tee_supplicant/src/ree_fs.c [hunk 7]
```diff
@@ -270,7 +324,33 @@ static int tee_fs_remove(size_t num_params, struct tee_param *params)
...
+		}
+		rc = fs_readdir(&dir, &entry);
+		fs_closedir(&dir);
+		if (rc < 0) {
+			return rc;
+		}
+		if (entry.name[0]) {
+			break;
+		}
+		rc = fs_unlink(path);
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: I am wondering what would happen in multithreaded environment, when there will be a chance that other CPU would create a new file between your `fs_readdirs` and `fs_unlink` calls.

File: tee_supplicant/src/ree_fs.c [hunk 8]
```diff
@@ -69,10 +63,10 @@ static char *dname(char *path)
 	return NULL;
 }
 
-static int tee_fs_open(size_t num_params, struct tee_param *params,
+static int tee_fs_open(struct thread_arg *arg, size_t num_params, struct tee_param *params,
 		       fs_mode_t flags)
 {
-	char *name, path[REE_FS_PATH_MAX] = REE_FS_MP;
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: Could you just replace `REE_FS_MP` with `CONFIG_OPTEE_STORAGE_ROOT` here and drop all those changes with thread parameter passing?

File: tee_supplicant/src/tee_supplicant.c [hunk 9]
```diff
@@ -324,6 +325,51 @@ static void tee_supp_main(void *p1, void *p2, void *p3)
 	}
 }
 
+static char *_strdup(const char *str)
+{
+	int len;
+	char *s;
+
+	len = strlen(str)+1;
+	s = k_malloc(len);
```
Comment (lorc) [inline tee_supplicant/src/tee_supplicant.c:330]: `k_malloc` can fail

File: tee_supplicant/src/tee_supplicant.c [hunk 10]
```diff
@@ -340,9 +346,20 @@ static int tee_supp_init(const struct device *dev)
...
 		return -EINVAL;
 	}
+	arg = k_malloc(sizeof(struct thread_arg));
+	if (!arg) {
+		LOG_ERR("Unable to allocate thread arg");
+		return -ENOMEM;
+	}
+	arg->tee_fs_root = CONFIG_OPTEE_STORAGE_ROOT;
+
+	rc = mkpath(CONFIG_OPTEE_STORAGE_ROOT);
```
Comment (lorc) [inline tee_supplicant/src/tee_supplicant.c:396]: You are using `mkpath` here but introduce it in the next patch only

File: tee_supplicant/src/tee_supplicant.c [hunk 11]
```diff
@@ -324,6 +325,51 @@ static void tee_supp_main(void *p1, void *p2, void *p3)
...
+}
+
+static int mkpath(const char *path)
+{
+	int status = 0;
+	char *subpath = _strdup(path);
+	char *prev = subpath;
+	char *curr = NULL;
+
+	while (status == 0 && (curr = strchr(prev, '/')) != 0) {
```
Comment (lorc) [inline tee_supplicant/src/tee_supplicant.c:359]: Again, there can be weird cases with paths like `////` or `/../../`. Now, when base TEE storage path is user-configurable, there will be someone who'll that option to `/tee//`.

File: tee_supplicant/src/ree_fs.c [hunk 12]
```diff
@@ -195,16 +195,24 @@ static int tee_fs_read(size_t num_params, struct tee_param *params)
...
-		return TEEC_ERROR_ITEM_NOT_FOUND;
-	}
-
-	sz = fs_read(file, buf, len);
-	if (sz < 0) {
-		LOG_ERR("read failure (%ld)", sz);
-		return TEEC_ERROR_GENERIC;
+	sz = 0;
+	while (len > 0) {
+		rc = fs_seek(file, offset+sz, SEEK_SET);
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: Formatting: `offset + sz`

Comment (lorc) [reply tee_supplicant/src/ree_fs.c]: Also, why are you doing seek on every iteration? IMO, you need to make it only once and then call `fs_read` inside the loop

File: tee_supplicant/src/ree_fs.c [hunk 13]
```diff
@@ -244,16 +252,25 @@ static int tee_fs_write(size_t num_params, struct tee_param *params)
...
-		LOG_ERR("invalid offset %ld (%d)", offset, rc);
-		return TEEC_ERROR_ITEM_NOT_FOUND;
-	}
 
-	sz = fs_write(file, buf, len);
-	if (sz < 0) {
-		LOG_ERR("write failure (%ld)", sz);
-		return TEEC_ERROR_GENERIC;
+	sz = 0;
+	while (len > 0) {
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:262]: The same comments as for the `read` function

**Review by lorc — CHANGES_REQUESTED** [review]

File: tee_supplicant/src/tee_supplicant.c (hunk 10 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c:396]: Fixed

File: tee_supplicant/src/tee_supplicant.c (hunk 9 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c:330]: Fixed

File: tee_supplicant/src/ree_fs.c (hunk 4 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: I guess those cases should be processed by file system. Isn't it?

File: tee_supplicant/src/ree_fs.c (hunk 2 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: As far as I can understand you are talking about some comments removal.
Fixed

File: tee_supplicant/src/ree_fs.c (hunk 5 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:153]: Technically This is obsolete. Handled in https://github.com/xen-troops/zephyr-optee-client/pull/8/commits/ea68d9949f4747cd17f1b16f2a560395da92a160

File: tee_supplicant/src/ree_fs.c (hunk 6 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:203]: Technically This is obsolete. Handled in https://github.com/xen-troops/zephyr-optee-client/pull/8/commits/ea68d9949f4747cd17f1b16f2a560395da92a160

File: tee_supplicant/src/tee_supplicant.c (hunk 11 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c:359]: There is file system driver's responsibility to handle those cases

File: tee_supplicant/src/ree_fs.c (hunk 12 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Right. Fixed

File: tee_supplicant/src/ree_fs.c (hunk 13 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:262]: Fixed

File: tee_supplicant/src/ree_fs.c (hunk 3 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Added file object registry

File: tee_supplicant/src/ree_fs.c (hunk 2 above)
Comment (lorc) [reply tee_supplicant/src/ree_fs.c]: Not quite. You are moving big chunks of the code and this is not covered in the commit message. So it is hard to me to understand if your changes are correct.

File: tee_supplicant/src/ree_fs.c (hunk 3 above)
Comment (lorc) [reply tee_supplicant/src/ree_fs.c]: I didn't quite get this. Could you please elaborate?

File: tee_supplicant/src/ree_fs.c (hunk 4 above)
Comment (lorc) [reply tee_supplicant/src/ree_fs.c]: Well, it depends on how you use this `dname` function

File: tee_supplicant/src/tee_supplicant.c (hunk 11 above)
Comment (lorc) [reply tee_supplicant/src/tee_supplicant.c:359]: Have you tested for such cases?

File: tee_supplicant/src/ree_fs.c [hunk 14]
```diff
@@ -15,18 +15,65 @@
 
 LOG_MODULE_REGISTER(ree_fs);
 
-static int internal_open(char *path, fs_mode_t flags)
+#define MAX_FILES		20
+
+static struct fs_file_t *files[MAX_FILES] = {0};
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:23]: Do you really need all this? While OP-TEE secure storage protocol uses term "fd", it actually means "a 64-bit number".  Instead of creating this storage to map short ints to pointers, you can just pass pointer to `fs_file_t` as `fd` and it should work with no problem. 

If you want to go with this "registry" path, you at least need to provide a proper locking.

Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:23]: There is can be problems with error cases. Error case is negative value and pointer (on 64-bit system) can be negative

File: tee_supplicant/src/ree_fs.c (hunk 3 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: This fixed in https://github.com/xen-troops/zephyr-optee-client/pull/8/commits/49fbd30d8649e8166c620bff61e79594b634f418

File: tee_supplicant/src/tee_supplicant.c (hunk 11 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c:359]: Yes. I have tested.

File: tee_supplicant/src/ree_fs.c (hunk 14 above)
Comment (lorc) [reply tee_supplicant/src/ree_fs.c:23]: Okay, but what about locking? This may and will be used in the multithreaded environment.

**Review by lorc — CHANGES_REQUESTED** [review]
Could you please squash fixups into appropriate commits and remove unrelated changes. For example, you are introducing use of `z_alloc_fd` and then removing it back. What is the point in this? It only makes it harder to review the changes.

File: tee_supplicant/src/ree_fs.c (hunk 14 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:23]: I use locking here. Line 21 declares mutex

File: tee_supplicant/src/tee_supplicant.c [hunk 15]
```diff
@@ -377,6 +377,7 @@ static int tee_supp_init(const struct device *dev)
 	struct tee_version_info info = { 0 };
 	int rc;
 
+	LOG_ERR("supplicant initialized");
```
Comment (lorc) [inline tee_supplicant/src/tee_supplicant.c]: Your commit message says "Updated copyright message" but you also making changes to the logging

**Review by lorc — CHANGES_REQUESTED** [review]
May I ask you to limit this PR only to functional changes? 

It is already has 8 commits to review, we don't need additional changes that are not related to this PR.

File: tee_supplicant/src/tee_supplicant.c (hunk 15 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Sorry. I missed this. Removed

**Comment by dsemenets** [comment]
> Could you please squash fixups into appropriate commits and remove unrelated changes. For example, you are introducing use of `z_alloc_fd` and then removing it back. What is the point in this? It only makes it harder to review the changes.

I didn't introduce z_alloc_fd. It was introduced in initial version of ree_fs. I just reused this solution.

File: tee_supplicant/src/ree_fs.c (hunk 4 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: I have updated dname function

File: tee_supplicant/src/ree_fs.c (hunk 2 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: I have redesigned tee_fs_open function to support directory creation if it needed.

**Comment by lorc** [comment]
> > Could you please squash fixups into appropriate commits and remove unrelated changes. For example, you are introducing use of z_alloc_fd and then removing it back. What is the point in this? It only makes it harder to review the changes.

> I didn't introduce z_alloc_fd. It was introduced in initial version of ree_fs. I just reused this solution.

Anyways, in "tee_supplicant: Supprort directory manipulation on file operations" you introduce `internal_open` function that adds use of `z_alloc_fd` then in "tee_supplicant: File object registry" you completely rework this code, by introducing `new_file` that removes `z_alloc_fd` call in favor for internal implementation. I don't see logic in this. Why do we adding code that is being removed in the next commit? Would it be better to introduce file object registry mechanism before reworking file and directory manipulation?

**Comment by dsemenets** [comment]
> > > Could you please squash fixups into appropriate commits and remove unrelated changes. For example, you are introducing use of z_alloc_fd and then removing it back. What is the point in this? It only makes it harder to review the changes.
> 
> > I didn't introduce z_alloc_fd. It was introduced in initial version of ree_fs. I just reused this solution.
> 
> Anyways, in "tee_supplicant: Supprort directory manipulation on file operations" you introduce `internal_open` function that adds use of `z_alloc_fd` then in "tee_supplicant: File object registry" you completely rework this code, by introducing `new_file` that removes `z_alloc_fd` call in favor for internal implementation. I don't see logic in this. Why do we adding code that is being removed in the next commit? Would it be better to introduce file object registry mechanism before reworking file and directory manipulation?

I have redesigned this part. Please look

File: tee_supplicant/src/ree_fs.c [hunk 16]
```diff
@@ -21,6 +20,68 @@ LOG_MODULE_REGISTER(ree_fs);
...
+	for (fd = 0; fd < MAX_FILES; fd++) {
+		if (!files[fd]) {
+			files[fd] = *file;
+			k_mutex_unlock(&file_mutex);
+			return fd;
+		}
+	}
+	k_mutex_unlock(&file_mutex);
+
+	return -EMFILE;
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:81]: Looks like you have a memory leak right here. I can't see code that frees `file`

File: tee_supplicant/src/ree_fs.c [hunk 17]
```diff
@@ -21,6 +20,68 @@ LOG_MODULE_REGISTER(ree_fs);
...
+		return NULL;
+	}
+
+	k_mutex_lock(&file_mutex, K_FOREVER);
+	file = files[fd];
+	k_mutex_unlock(&file_mutex);
+	return file;
+}
+
+static void remove_file(int fd)
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: `remove_file` sounds a bit misleading as it implies that file is being removed from a file system, when in fact you just close it and dealloc internal `fd`

File: tee_supplicant/src/ree_fs.c [hunk 18]
```diff
@@ -21,6 +20,68 @@ LOG_MODULE_REGISTER(ree_fs);
...
+	k_mutex_lock(&file_mutex, K_FOREVER);
+	file = files[fd];
+	if (file) {
+		k_free(file);
+	}
+	files[fd] = NULL;
+	k_mutex_unlock(&file_mutex);
+}
+
+static int new_file(struct fs_file_t **file)
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: `new_file` sounds a bit misleading as it implies that file is being created on a file system, when in fact you just crase `file` struct and allocate internal `fd`

File: tee_supplicant/src/ree_fs.c [hunk 19]
```diff
@@ -1,10 +1,9 @@
-// SPDX-License-Identifier: Apache-2.0
 /*
- * Copyright (c) 2023 EPAM Systems
+ * Copyright (c) 2024 EPAM Systems
  *
+ * SPDX-License-Identifier: Apache-2.0
  */
 
-
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:7]: Spurios change?

File: tee_supplicant/src/ree_fs.c [hunk 20]
```diff
@@ -1,10 +1,9 @@
-// SPDX-License-Identifier: Apache-2.0
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c:1]: Do you really need to move SPDX tag? AFAIK, it should be at the beginning of the file in a separate comment block.

And of course this should be done in a separate commit.

File: tee_supplicant/src/ree_fs.c [hunk 21]
```diff
@@ -82,12 +82,59 @@ static int new_file(struct fs_file_t **file)
...
+	}
+	return fd;
+free:
+	if (fd >= 0) {
+		remove_file(fd);
+	}
+	return rc;
+}
+
+static char *dname(char *path)
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: What paths should handle this function? My tests show couple of strange results. Could you please comment on them?

```
Test: '/a/' result: '(null)'
Test: '/a///' result: '/a'
Test: '' result: '(null)'
Test: '/a/b/' result: '/a'
Test: 'a///b' result: 'a//'
Test: 'a/./b' result: 'a/.'
Test: 'a/./b/' result: 'a/.'
Test: 'a/../b' result: 'a/..'
Test: 'a/../b/' result: 'a/..'
Test: 'x/..' result: 'x'
```

File: tee_supplicant/src/ree_fs.c [hunk 22]
```diff
@@ -108,30 +155,46 @@ static int tee_fs_open(size_t num_params, struct tee_param *params,
...
-			goto free;
+		if (flags & FS_O_CREATE) {
+			char *dir;
+
+			dir = dname(path);
+			if (!dir) {
+				rc = TEEC_ERROR_GENERIC;
+				goto out;
+			}
+			rc = fs_mkdir(dir);
```
Comment (lorc) [inline tee_supplicant/src/ree_fs.c]: It this function recursive? E.g. will it create both "tee" and "1" for path "/tee/1/abc"?

File: tee_supplicant/src/tee_supplicant.c [hunk 23]
```diff
@@ -320,10 +321,61 @@ static void tee_supp_main(void *p1, void *p2, void *p3)
...
+	}
+	return s;
+}
+
+static int do_mkdir(const char *path)
+{
+	int rc;
+
+	rc = fs_mkdir(path);
+	if (rc == 0 || rc == -EEXIST) {
```
Comment (lorc) [inline tee_supplicant/src/tee_supplicant.c]: No need for `rc == 0` check

**Review by lorc — CHANGES_REQUESTED** [review]
Also please fix spelling:

"tee_supplicant: Add **possibility** **for** manual supplicant initialization"

" Add option how supplicant initialization should **be** done: "

" Amount of actually processed bytes during read/write operation can
    be less **than** amount of requested bytes."

"tee_supplicant: **Support** directory manipulation on file operations"

There may be more mistakes, please run spellchecker.

File: tee_supplicant/src/ree_fs.c (hunk 16 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:81]: Fixed

File: tee_supplicant/src/ree_fs.c (hunk 17 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Renamed

File: tee_supplicant/src/ree_fs.c (hunk 18 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Renamed

File: tee_supplicant/src/ree_fs.c (hunk 20 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c:1]: fixed

File: tee_supplicant/src/tee_supplicant.c (hunk 23 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Fixed

File: tee_supplicant/src/ree_fs.c (hunk 21 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Investigation shows OPTEE doesn't ask to create file in sub directory. No need create directory. This patch removed

File: tee_supplicant/src/ree_fs.c (hunk 8 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: Fixed

File: tee_supplicant/src/ree_fs.c (hunk 22 above)
Comment (dsemenets) [reply tee_supplicant/src/ree_fs.c]: This patch removed

**Review by lorc — APPROVED** [review]
Thanks

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
